File: node-bindings/classic-node/__test__/config_operations_conformance.ts
import {mkdtemp, readdir, readFile, rm, writeFile} from "node:fs/promises";
import {tmpdir} from "node:os";
import {join, relative, sep} from "node:path";
import * as classic from "../index.js";
import {yamlValues} from "./config_yaml_values_conformance.js";

type JsonObject = Record<string, any>;

/** Re-read every durable file, retaining exact UTF-8 bytes and portable paths. */
async function inventory(root: string, directory = root): Promise<JsonObject[]> {
    const result: JsonObject[] = [];
    for (const entry of await readdir(directory, {withFileTypes: true})) {
        const path = join(directory, entry.name);
        if (entry.isSymbolicLink()) throw new Error("unexpected link in config workspace");
        if (entry.isDirectory()) result.push(...await inventory(root, path));
        else result.push({path: relative(root, path).split(sep).join("/"), content: await readFile(path, "utf8")});
    }
    return result.sort((a, b) => a.path.localeCompare(b.path));
}

/** Invoke the public explicit loader, retaining structured failures and all durable effects. */
export async function observeConfigOperations(fixture: JsonObject): Promise<JsonObject> {
    if (!["load-explicit", "main-version", "persist-local"].includes(fixture.operation) || Object.keys(fixture).sort().join() !== (fixture.operation === "persist-local" ? "docsRoot,files,gameRoot,operation" : "files,operation")) {
        throw new Error("unsupported config operation fixture");
    }
    const root = await mkdtemp(join(tmpdir(), "classic-config-conformance-"));
    try {
        for (const [name, content] of Object.entries(fixture.files)) {
            if (!["main.yaml", "game.yaml", "ignore.yaml", "CLASSIC Main.yaml", "local.yaml", "CLASSIC Settings.yaml"].includes(name) || typeof content !== "string") {
                throw new Error("config fixture requires owned YAML filenames and UTF-8 text");
            }
            await writeFile(join(root, name), content, "utf8");
        }
        if (fixture.operation === "persist-local") {
            await classic.persistGameLocalPaths(join(root, "local.yaml"), fixture.gameRoot, fixture.docsRoot);
            return {result: null, error: null, files: await inventory(root)};
        }
        if (fixture.operation === "main-version") {
            // Own resolver inputs while this sequential adapter awaits native loading.
            const previous = new Map(["LOCALAPPDATA", "XDG_CACHE_HOME"].map(name => [name, process.env[name]]));
            try {
                for (const name of previous.keys()) process.env[name] = join(root, "isolated-cache");
                const version = await classic.loadMainYamlVersion(root);
                return {result: {version}, error: null, files: await inventory(root)};
            } finally {
                for (const [name, value] of previous) {
                    if (value === undefined) delete process.env[name]; else process.env[name] = value;
                }
            }
        }
        let result: JsonObject | null = null;
        let error: JsonObject | null = null;
        let snapshot: classic.ExplicitYamlDataSnapshot | undefined;
        try {
            snapshot = await classic.loadExplicitYamlData(
                {
                    mainPath: join(root, "main.yaml"),
                    gamePath: join(root, "game.yaml"),
                    ignorePath: join(root, "ignore.yaml")
                },
                classic.JsGameId.Fallout4,
                "auto",
            );
        } catch (failure: any) {
            // Reject infrastructure failures instead of manufacturing a domain observation.
            if (!["read", "parse", "invalid_utf8", "invalid_role_data", "unsupported_game"].includes(failure.code)) throw failure;
            error = {
                code: failure.code,
                role: failure.yamlRole ?? null,
                path: failure.path == null ? null : relative(root, failure.path).split(sep).join("/")
            };
        }
        if (snapshot !== undefined) {
            if (snapshot.game !== classic.JsGameId.Fallout4) throw new Error("explicit snapshot returned an unexpected game");
            const data = snapshot.yamlData;
            if (snapshot.gameDataRole !== classic.JsExplicitYamlDataGameRole.Fallout4) throw new Error("explicit snapshot returned an unexpected game role");
            result = {
                game: "Fallout4",
                classicVersion: data.classicVersion,
                xseAcronym: data.xseAcronym,
                crashgenName: data.crashgenName,
                gameVersion: data.gameVersion,
                ignoreList: data.ignoreList
            };
            result.gameRole = "Fallout4";
            result.yamlValues = yamlValues(data);
            const fromContent = classic.createYamlDataFromContent(fixture.files["main.yaml"], fixture.files["game.yaml"], fixture.files["ignore.yaml"], "Fallout4", "auto");
            if (JSON.stringify(yamlValues(fromContent)) !== JSON.stringify(result.yamlValues)) throw new Error("content constructor differs from retained explicit data");
            result.identities = Object.fromEntries([["main.yaml", snapshot.mainIdentity], ["game.yaml", snapshot.gameIdentity], ["ignore.yaml", snapshot.ignoreIdentity]].map(([name, identity]: any) => [name, {
                sha256: identity.sha256,
                byteLen: Number(identity.byteLen)
            }]));
        }
        return {result, error, files: await inventory(root)};
    } finally {
        await rm(root, {recursive: true, force: true});
    }
}

File: python-bindings/tests/config_operations_conformance.py
"""Observe explicit config loading through real Python bindings and owned files."""

from __future__ import annotations

import os
import tempfile
from collections.abc import Mapping
from pathlib import Path
from typing import Any


def observe_config_operations(fixture: Mapping[str, Any]) -> dict[str, Any]:
    """Load only caller-selected files and re-read all durable bytes after execution.

    Domain errors retain their structured role and path. Unexpected adapter errors
    propagate, and temporary storage is removed even when native execution fails.
    """
    import classic_config
    from config_yaml_values_conformance import yaml_values

    if fixture.get("operation") not in {
        "load-explicit",
        "main-version",
        "persist-local",
        "clear-cache",
    } or set(fixture) != {
        "operation",
        "files",
    } | (
            {"gameRoot", "docsRoot"}
            if fixture.get("operation") == "persist-local"
            else set()
    ):
        raise ValueError("unsupported config operation fixture")
    with tempfile.TemporaryDirectory(prefix="classic-config-conformance-") as directory:
        root = Path(directory)
        for name, content in fixture["files"].items():
            if name not in {
                "main.yaml",
                "game.yaml",
                "ignore.yaml",
                "CLASSIC Main.yaml",
                "local.yaml",
                "CLASSIC Settings.yaml",
            } or not isinstance(content, str):
                raise ValueError(
                    "config fixture requires owned YAML filenames and UTF-8 text"
                )
            (root / name).write_bytes(content.encode("utf-8"))
        if fixture["operation"] == "clear-cache":
            # The extension owns its own Rust statics. This boundary proves the
            # public void return and read-only disk contract, not hidden eviction.
            first = classic_config.clear_yaml_cache()
            second = classic_config.clear_yaml_cache()
            return {
                "result": [first, second],
                "error": None,
                "files": [
                    {"path": path.name, "content": path.read_bytes().decode("utf-8")}
                    for path in sorted(root.iterdir())
                    if path.is_file()
                ],
            }
        if fixture["operation"] == "persist-local":
            classic_config.persist_game_local_paths(
                root / "local.yaml", fixture["gameRoot"], fixture["docsRoot"]
            )
            return {
                "result": None,
                "error": None,
                "files": [
                    {"path": path.name, "content": path.read_bytes().decode("utf-8")}
                    for path in sorted(root.iterdir())
                    if path.is_file()
                ],
            }
        if fixture["operation"] == "main-version":
            # The public loader resolves process cache roots; cases run sequentially.
            previous = {
                name: os.environ.get(name)
                for name in ("LOCALAPPDATA", "XDG_CACHE_HOME")
            }
            try:
                for name in previous:
                    os.environ[name] = str(root / "isolated-cache")
                version = classic_config.load_main_yaml_version(str(root))
            finally:
                for name, value in previous.items():
                    if value is None:
                        os.environ.pop(name, None)
                    else:
                        os.environ[name] = value
            return {
                "result": {"version": version},
                "error": None,
                "files": [
                    {"path": path.name, "content": path.read_bytes().decode("utf-8")}
                    for path in sorted(root.iterdir())
                    if path.is_file()
                ],
            }
        result = error = None
        try:
            snapshot = classic_config.load_explicit_yaml_data(
                classic_config.ExplicitYamlDataPaths(
                    root / "main.yaml", root / "game.yaml", root / "ignore.yaml"
                ),
                classic_config.ExplicitYamlDataGame.FALLOUT4,
                "auto",
            )
        except classic_config.ExplicitYamlDataLoadError as failure:
            error = {
                "code": failure.code,
                "role": failure.yaml_role,
                "path": None
                if failure.path is None
                else Path(failure.path).relative_to(root).as_posix(),
            }
        else:
            if snapshot.game != classic_config.ExplicitYamlDataGame.FALLOUT4:
                raise ValueError("explicit snapshot returned an unexpected game")
            data = snapshot.yaml_data
            values = yaml_values(data)
            from_content = classic_config.YamlData.from_yaml_content(
                fixture["files"]["main.yaml"],
                fixture["files"]["game.yaml"],
                fixture["files"]["ignore.yaml"],
                "Fallout4",
                "auto",
            )
            if yaml_values(from_content) != values:
                raise ValueError(
                    "content constructor differs from retained explicit data"
                )
            if (
                    repr(data)
                    != f"YamlData(game={data.crashgen_name.split('_')[0]}, version={data.classic_version})"
            ):
                raise ValueError(
                    "public YAML representation lost its identifying fields"
                )
            result = {
                "yamlValues": values,
                "game": "Fallout4",
                "gameRole": snapshot.game_data_role,
                "identities": {
                    name: {"sha256": value.sha256, "byteLen": value.byte_len}
                    for name, value in (
                        ("main.yaml", snapshot.main_identity),
                        ("game.yaml", snapshot.game_identity),
                        ("ignore.yaml", snapshot.ignore_identity),
                    )
                },
                "classicVersion": data.classic_version,
                "xseAcronym": data.xse_acronym,
                "crashgenName": data.crashgen_name,
                "gameVersion": data.game_version,
                "ignoreList": list(data.ignore_list),
            }
        files = [
            {
                "path": path.relative_to(root).as_posix(),
                "content": path.read_bytes().decode("utf-8"),
            }
            for path in sorted(root.rglob("*"))
            if path.is_file()
        ]
        return {"result": result, "error": error, "files": files}
